### PR TITLE
Simple PageFile format

### PR DESCRIPTION
**Pull request by viczhang861**
There is no split support yet. All pages are written continuously to a single file.

Existing tests in `HiveIntegrationSmokeTes` and `TestHivePageSink` cover new formats.

Attached benchmark result shows much faster read/write compared with ORC format. Reading from PageFile format is 4X faster; Writing PageFile format is 40X faster.

91361137_564002691179682_3632670565724061696_n

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageDataSink.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Gets the size of the memory buffers.
+     */
+    long getRetainedSizeInBytes();
+
+    /**
+     * Write a page
+     */
+    void write(Page dataPage)
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageDataSink.java:35]: From what I understand this interface is added by analogy to the `OrcDataSink`.

The main difference that I see right now is that the `OrcDataSink` accepts the data for the entire block in a single call. In other words the buffering of the block is happening before calling the `OrcDataSink#write`. The buffering happens in the OrcWriter. 

From one side it leads to somehow tricky and not very obvious "side effects" or implications of the ORC stripe size. By changing the stripe size we are effectively changing the block size of the underlying file system. Thus in some sense it feels like this logic must be pushed down to the `PageDataSink` interface, so the buffering and block size configuration happens there. But then I realize that if we want to add the meta-information to make the format splittable, we don't wanna add check points in a middle of a block. And if the block buffering is hidden behind the `PageDataSink` interface there will be no way to know where are the block breaking point.

Based on this thoughts it feels like we should stick with the existing ORC model, when the entire block is supplied. We can make this method to accept a `List<Page>`.

Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageDataSink.java:35]: Also it feels like this method should accept `SerializedPage`. The `OrcDataSink` interface accepts `OrcDataOutput`, that is effective a "lazy" serialized representation. I don't think that we need the laziness in our case though.

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/OutputStreamPageDataSink.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        writeSerializedPage(output, pagesSerde.serialize(dataPage));
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        output.flush();
+        output.close();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/OutputStreamPageDataSink.java:64]: Doesn't it flush automatically on close?

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     * Number of bytes already flushed to the data sink.
+     */
+    public long getWrittenBytes()
+    {
+        return pageDataSink.size();
+    }
+
+    public long getRetainedBytes()
+    {
+        return INSTANCE_SIZE + pageDataSink.getRetainedSizeInBytes();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:47]: This method is unused

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import static java.util.Objects.requireNonNull;
+
+public class PageWriter
+        implements Closeable
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(PageWriter.class).instanceSize();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:27]: This is effectively unused

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.concurrent.Callable;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_DATA_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class PageFileWriter
+        implements HiveFileWriter
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(PageFileWriter.class).instanceSize();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java:32]: This is unused

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    private final PageWriter pageWriter;
+    private final Callable<Void> rollbackAction;
+
+    private long validationCpuNanos;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java:37]: This is unused

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    @Override
+    public long getWrittenBytes()
+    {
+        return pageWriter.getWrittenBytes();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java:56]: Should this be implemented?

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriterFactory.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                return null;
+            };
+            return Optional.of(new PageFileWriter(pageDataSink, rollbackAction));
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_WRITER_OPEN_ERROR, "Error creating pagefile", e);
+        }
+    }
+
+    protected PageDataSink createPageDataSink(FileSystem fileSystem, Path path)
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriterFactory.java:88]: note: Generally having a composition based factory is preferred. However since the ORC writer is already implemented this way - let's keep it so for consistency.

File: presto-hive/src/test/java/com/facebook/presto/hive/benchmark/FileFormat.java [hunk 9]
```diff
@@ -202,6 +207,36 @@ public boolean supportsDate()
...
+                HiveCompressionCodec compressionCodec)
+                throws IOException
+        {
+            return new PrestoPageFormatWriter(targetFile);
+        }
+
+        @Override
+        public boolean supportsDate()
+        {
+            return false;
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/benchmark/FileFormat.java:236]: It should support date

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java [hunk 10]
```diff
@@ -1278,8 +1280,8 @@ public void testMismatchSchemaTable()
         boolean pushdownFilterEnabled = getHiveClientConfig().isPushdownFilterEnabled();
 
         for (HiveStorageFormat storageFormat : createTableFormats) {
-            // TODO: fix coercion for JSON
-            if (storageFormat == JSON) {
+            // TODO: fix coercion for JSON or PAGEFILE
+            if (storageFormat == JSON || storageFormat == PAGEFILE) {
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:1284]: It should be pretty simple to implement schema evolution. However since it is not needed I aggree we shouldn't be doing this. The only recommendation I have is to add a specific error message (currently we are failing with ArrayOutOfBound)

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: 13
	at com.facebook.presto.spi.Page.getBlock(Page.java:103)
	at com.facebook.presto.hive.pagefile.PageFilePageSource.getNextPage(PageFilePageSource.java:99)
	at com.facebook.presto.hive.HivePageSource.getNextPage(HivePageSource.java:126)
	... 29 more

```

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePageSink.java [hunk 11]
```diff
@@ -122,6 +123,10 @@ public void testAllFormats()
                     if (codec == NONE || !codec.isSupportedStorageFormat(format)) {
                         continue;
                     }
+                    // No compression support needed for PAGEFILE
+                    if (format == PAGEFILE) {
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePageSink.java:127]: It is fine to do not add compression support in this PR. However we should add the compression. Writing temporary files without any compression might be too costly IOPS wise.

**Comment by mbasmanova** [comment]
CC: @oerling @sujay-jain

**Comment by yingsu00** [comment]
@viczhang861 Hi Vic, what's the motivation to introduce this new format? If it's performance improvement, could you add the JMH result to the PR? Thank you!

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveStorageFormat.java [hunk 12]
```diff
@@ -83,6 +83,11 @@
             LazySimpleSerDe.class.getName(),
             TextInputFormat.class.getName(),
             HiveIgnoreKeyTextOutputFormat.class.getName(),
+            new DataSize(8, Unit.MEGABYTE)),
+    PAGEFILE(
+            LazySimpleSerDe.class.getName(),
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveStorageFormat.java]: Changed to empty string to indicate SerDe is not applicable.

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageDataSink.java (hunk 1 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageDataSink.java]: Changed it to `void write(Iterable<SerializedPage> pages)` ,   the remaining work is to write footer.  After some deeper investigation,  `OrcDataOutput` is mainly designed for multiple streams from all column writers,  PageFile would have only one `outputstream`,  l plan to wrap footer into a slice and not mix it with `SerializedPage`.

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/OutputStreamPageDataSink.java (hunk 2 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/OutputStreamPageDataSink.java:60]: Good catch, thank you !

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java (hunk 10 above)
Comment (viczhang861) [reply presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:1284]: Index check is added.

**Comment by viczhang861** [comment]
> @viczhang861 Hi Vic, what's the motivation to introduce this new format? If it's performance improvement, could you add the JMH result to the PR? Thank you!

To materialize intermediate results faster, ORC is not needed. I attached some preliminary results by comparing Page format with ORC and demonstrates faster read/write.

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSource.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        long start = System.nanoTime();
+
+        Page page = pageReader.next();
+
+        Block[] blocks = new Block[hiveColumnIndexes.length];
+        for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
+            if (hiveColumnIndexes[fieldId] >= page.getChannelCount()) {
+                throw new PrestoException(
+                        NOT_SUPPORTED,
+                        "schema mismatched table is not supported for PageFile format");
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSource.java:104]: We call it `schema evolution`. Let's change it to `schema evolution is not supported for page file format`

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSourceFactory.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        if (!PageInputFormat.class.getSimpleName().equals(storage.getStorageFormat().getInputFormat())) {
+            return Optional.empty();
+        }
+
+        FSDataInputStream inputStream;
+        try {
+            inputStream = hdfsEnvironment.getFileSystem(session.getUser(), path, configuration).openFile(path, hiveFileContext);
+        }
+        catch (Exception e) {
+            if (nullToEmpty(e.getMessage()).trim().equals("Filesystem closed") ||
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSourceFactory.java:87]: I'm assuming this logic is transferred here from the Orc format

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            catch (Exception ignored) {
+                // ignore
+            }
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error committing write to Hive", e);
+        }
+    }
+
+    @Override
+    public void rollback()
+    {
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java:92]: Should the rollback action be run here as well?

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+    private final OrcDataSink dataSink;
+    private long bufferedBytes;
+    private long retainedBytes;
+    private long maxBufferedBytes = (new DataSize(128, MEGABYTE)).toBytes();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:38]: Remove `(` and `)`. Ideally we should also  make it configurable (with a session and configuration property, similarly to ORC).

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    private List<OrcDataOutput> bufferedPages;
+
+    public PageWriter(OrcDataSink dataSink)
+    {
+        this.dataSink = requireNonNull(dataSink, "pageDataSink is null");
+        bufferedPages = new ArrayList<>();
+    }
+
+    private boolean closed;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:47]: let's move it to the other fields (before the constructor)

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    public void write(SerializedPage page)
+            throws IOException
+    {
+        retainedBytes += page.getRetainedSizeInBytes();
+        PageDataOutput pageDataOutput = new PageDataOutput(page);
+        long writtenSize = pageDataOutput.size();
+        if (maxBufferedBytes - bufferedBytes < writtenSize) {
+            dataSink.write(bufferedPages);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:63]: Isn't the current page being lost?

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+            bufferedPages.clear();
+            bufferedBytes = 0;
+            retainedBytes = 0;
+        }
+        else {
+            bufferedPages.add(new PageDataOutput(page));
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:69]: Why to recreate it if it is already created?

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java (hunk 18 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:63]: Given that this bug wasn't caught by test I would suggest improving the test (or adding a unit test) to cover this case.

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSource.java [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+    private long memoryUsageBytes;
+
+    public PageFilePageSource(
+            FSDataInputStream inputStream,
+            PagesSerde pagesSerde,
+            List<HiveColumnHandle> columns)
+    {
+        this.inputStream = requireNonNull(inputStream, "inputStream is null");
+        pageReader = readPages(
+                requireNonNull(pagesSerde, "pagesSerdeis null"),
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageSource.java:53]: `pageSerde is null`

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java (hunk 15 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java:92]: It looks like the `OrcFileWriter` does both, it closes the output file by closing the `pageWriter` and also runs the rollback action: https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriter.java#L176

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageInputFormat.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapred.FileInputFormat;
+import org.apache.hadoop.mapred.InputSplit;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.RecordReader;
+import org.apache.hadoop.mapred.Reporter;
+
+public class PageInputFormat
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageInputFormat.java:25]: Could you please remind me why this class is needed?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 22]
```diff
@@ -311,6 +312,10 @@ public static void setReadColumns(Configuration configuration, List<Integer> rea
             return MapredParquetInputFormat.class;
         }
 
+        if (PageInputFormat.class.getSimpleName().equals(inputFormatName)) {
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:315]: Could you please remind me why this change is needed?

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+            bufferedBytes += writtenSize;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:70]: Why do we need both, buffered and retained?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 22 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:315]: > Could you please remind me why this change is needed?

Otherwise, the line after this will fail :  `Class<?> clazz = conf.getClassByName(inputFormatName);`

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageInputFormat.java (hunk 21 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageInputFormat.java:25]: `BackgroundHiveSplitLoader` requires a class `InputFormat<?, ?> inputFormat = getInputFormat(configuration, inputFormatName, false);`

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java (hunk 23 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:70]: `retainedSize` is used for memory tracking,  `bufferedBytes` is to track the size to be written to file

File: presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java (hunk 18 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageWriter.java:62]: Thank you ! Session property and test is added to test no buffer of page inside `PageWriter` (write page by page)
